pkg/controller/v1alpha1/localmodelnode/controller.go

```diff
@@ -45,13 +55,237 @@ type LocalModelNodeReconciler struct {
+					Volumes: []v1.Volume{
+						{
+							Name: "kserve-pvc-source",
+							VolumeSource: v1.VolumeSource{
+								PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+									ClaimName: claimName,
+								},
+							},
+						},
+					},
+					SecurityContext: &v1.PodSecurityContext{
+						FSGroup: FSGroup,
+					},
+				},
+			},
+		},
+	}
+	dryrunJob, err := jobs.Create(context.TODO(), expectedJob, metav1.CreateOptions{DryRun: []string{metav1.DryRunAll}})
+	if err != nil {
+		return nil, err
+	}
+	if job != nil && reflect.DeepEqual(job.Spec.Template.Spec, dryrunJob.Spec.Template.Spec) {
+		return job, nil
+	}
+
+	if jobFound {
+		bg := metav1.DeletePropagationBackground
+		err = jobs.Delete(context.TODO(), job.Name, metav1.DeleteOptions{
+			PropagationPolicy: &bg,
+		})
+		if err != nil {
+			c.Log.Error(err, "Failed to delete job.", "name", job.Name)
+			return nil, err
+		}
+	}
+
+	if err := controllerutil.SetControllerReference(localModelNode, expectedJob, c.Scheme); err != nil {
+		c.Log.Error(err, "Failed to set controller reference", "name", modelInfo.ModelName)
+		return nil, err
+	}
+	expectedJob.Name = jobName
+	job, err = jobs.Create(context.TODO(), expectedJob, metav1.CreateOptions{})
+	c.Log.Info("Creating job", "name", job.Name, "namespace", namespace)
+	if err != nil {
+		c.Log.Error(err, "Failed to create job.", "name", expectedJob.Name)
+		return nil, err
+	}
+	return job, err
+}
+
+// Fetches container spec for model download container, use the default KServe image if not found
+func (c *LocalModelNodeReconciler) getContainerSpecForStorageUri(storageUri string) (*v1.Container, error) {
+	storageContainers := &v1alpha1api.ClusterStorageContainerList{}
+	if err := c.Client.List(context.TODO(), storageContainers); err != nil {
+		return nil, err
+	}
+
+	for _, sc := range storageContainers.Items {
+		if sc.IsDisabled() {
+			continue
+		}
+		if sc.Spec.WorkloadType != v1alpha1api.LocalModelDownloadJob {
+			continue
+		}
+		supported, err := sc.Spec.IsStorageUriSupported(storageUri)
+		if err != nil {
+			return nil, fmt.Errorf("error checking storage container %s: %w", sc.Name, err)
+		}
+		if supported {
+			return &sc.Spec.Container, nil
+		}
+	}
+
+	defaultContainer := &v1.Container{
+		Image:                    defaultJobImage,
+		TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+	}
+	return defaultContainer, nil
+}
+
+func (c *LocalModelNodeReconciler) DownloadModel(ctx context.Context, localModelNode *v1alpha1api.LocalModelNode, jobNamespace string) error {
+	c.Log.Info("Downloading models to", "node", localModelNode.ObjectMeta.Name)
+
+	for _, modelInfo := range localModelNode.Spec.LocalModels {
+		jobName := modelInfo.ModelName + "-" + localModelNode.ObjectMeta.Name
+		c.Log.Info("Launch download job", "name", jobName)
+		job, err := c.launchDownloadJob(jobName, jobNamespace, localModelNode, modelInfo, modelInfo.ModelName, localModelNode.ObjectMeta.Name)
```

Comment:
@yuzisun 
In the old code, we launch a download job if no job exists AND model status != download successful (no job was successful before).
Launch a download job when
Option 1: if (no job exists AND model status != download successful) or model folder does not exist 
Option 2: if no job exists or model folder does not exist
Isn't option 2 better because it doesn't depend on it's own status field if we assume download jobs are never deleted.

we don't need to change anything in this pr.

Reply:
Another issue - when we deploy this change, our existing jobs are owned by the controller, not the agent. Need to test how it works. Worst case it would download all models again which is fine.

Reply:
I agree, with the node agent we should no longer depend on the download job status. I am not sure if it is safe to assume that we never delete the download jobs, we can check the local model node resource if the model has been successfully downloaded.